scripts/release/other_identifiers/retrieve_indirectIdentifiers_from_mart.pl

```diff
@@ -1,77 +1,156 @@
+    my $species_abbreviation = shift;
+    my $identifier = shift;
+    my $output_file = shift;
+    my $attempt_count = shift // 1;
+    my $max_attempts = 3;
+
+    my $logger = get_logger(__PACKAGE__);
+
+    my $results = run_wget_query($species_abbreviation, $identifier);
+
+    my $output = $results->{'output'};
+    my $completion_tag = qr/\[success\]\n/;
+    if ($output =~ /$completion_tag/) {
+        $output =~ s/$completion_tag//;
+        write_output_to_file($output, $output_file);
+    } else {
+        my $error_message = "$identifier for species $species_abbreviation: download of identifier file not complete";
+        if ($attempt_count < $max_attempts) {
+            $logger->warn("$error_message on attempt $attempt_count - retrying");
+            download_identifier_file($species_abbreviation, $identifier, $output_file, ++$attempt_count);
+        } else {
+            my $error = $results->{'error'};
+            $logger->error("$error_message due to $error - aborting");
+        }
+    }
 }
 
-my $registry = get_registry();
-IDENTIFIER:foreach my $identifier (get_identifiers($sp_mart_name)) {
-    next if $identifier =~ /chembl|clone_based|dbass|description|ottg|ottt|ottp|shares_cds|merops|mirbase|reactome/;
-    my $query = get_query($registry);
-
-    $query->setDataset($sp_mart_name . "_gene_ensembl");
-
-    $query->addAttribute("ensembl_gene_id");
-    $query->addAttribute("ensembl_transcript_id");
-    $query->addAttribute("ensembl_peptide_id");
-    my $error_occurred;
-    try {
-        $query->addAttribute($identifier);
-    } catch {
-        $logger->warn("could not add attribute $identifier for $sp_mart_name");
-        $error_occurred = 1;
+# Runs query to EnsEMBL Mart for species and identifier
+# Returns the output and error from the wget command as a hash reference
+sub run_wget_query {
+    my $species_abbreviation = shift;
+    my $identifier = shift;
+
+    my $logger = get_logger(__PACKAGE__);
+
+    my $wget_query = get_wget_query_for_identifier($species_abbreviation, $identifier);
+
+    $logger->info("$identifier for species $species_abbreviation: beginning download of identifier file");
+    my ($output, $error) = capture {
+        system $wget_query;
     };
-    next IDENTIFIER if $error_occurred;
-    
-    $query->formatter("TSV");
-
-    my $query_runner = get_query_runner();
-    $query_runner->execute($query);
-    open(my $fh, '>', "output/$sp_mart_name\_$identifier");
-    $query_runner->printResults($fh);
+
+    return {
+        output => $output,
+        error => $error
+    };
+}
+
+# Write contents of identifier mappings from EnsEMBL Mart to output file
+sub write_output_to_file {
+    my $logger = get_logger(__PACKAGE__);
+
+    my $output = shift // $logger->logconfess("Need output to write to identifier file");
+    my $output_file = shift // $logger->logconfess("Need name of output file");
+
+    open(my $fh, '>', $output_file);
+    print {$fh} $output;

```

Comment:
Why are curly braces used here?

Reply:
Curly braces are a Perl Critic thing - it suggests to wrap file handles to make it explicit that they are a file handle and not data being printed.

https://metacpan.org/pod/Perl::Critic::Policy::InputOutput::RequireBracedFileHandleWithPrint